### PR TITLE
Update Showcase Demo with PubNub Unity SDK v7

### PR DESCRIPTION
**Pull request by olivercarson-pubnub**
This PR is to update the showcase demo with the PubNub Unity SDK v7
-Removal of old PubNub Unity SDK (v6)
-Adding new Unity SDK Package (v7)
-Update base showcase game (including multiplayer sync) with new unity sdk implementation.

There are still a few issues with the showcase:

1. As mentioned last time, the friend list online status thing is still not working right, but will be fixed in [DEVX-3359](https://pubnub.atlassian.net/browse/DEVX-3359)
2. Sometimes presence is not correctly updating, or changing suddenly when going into a lobby with two players in it. I've spent a good deal of time troubleshooting this, as it's not always consistent. I do have an item [DEVX-3352](https://pubnub.atlassian.net/browse/DEVX-3352) that covers Presence, so I didn't want to troubleshoot for too long in this change.
3. There will be a pop-up when you try and start the game with the new sdk relating to ScriptTemplates, which are just some new objects to create for the new sdk. I decided to not check those files in, as they will be generated when you start the editor anyways.
4. When repeating games (Start game ->main menu -> Lobby -> playing the game -> back to main menu -> lobby -> playing the game), sometimes there is a pubnub null object reference error getting thrown, as it's attempting to unsubscribe from all channels for the pubnub object that doesn't exist.

### CODE REVIEW

File: Assets/SuperMultiplayerShooter/Scripts/PNManager.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+public class PNManager : PNManagerBehaviour
+{
+    // UserId identifies this client.
+    public string userId;
+
+    //Persist the PubNub object across scenes
+    public static PNManager pubnubInstance;
+
+    //Cached players from connection.
+    private static Dictionary<string, UserMetadata> cachedPlayers = new Dictionary<string, UserMetadata>();
```
Comment (darryncampbell-pubnub) [inline Assets/SuperMultiplayerShooter/Scripts/PNManager.cs:15]: FYI, I'll need an alternative to PhotonInstance.Name (I think it is, basically, the nickname the player has assigned themselves).  I plan on using this cachedPlayers dictionary.

File: Assets/SuperMultiplayerShooter/Scripts/PNManager.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            Debug.LogError("Please set your PubNub keys in the PNConfigAsset.");
+        }
+        //Randomly generates a username. SystemInfo.deviceUniqueIdentifier does not work on WebGL Builds.
+        //string uuid = "User#" + Random.Range(0, 9999).ToString();
+        string uuid = System.Guid.NewGuid().ToString();
+
+        //Check if the user id already exists on this device. If not, save it.
+        if (PlayerPrefs.HasKey("uuid"))
+        {
+            uuid = PlayerPrefs.GetString("uuid");
```
Comment (darryncampbell-pubnub) [inline Assets/SuperMultiplayerShooter/Scripts/PNManager.cs:45]: I also make use of this PlayerPrefs UUID so I have a unique identifier for the player instance

File: Assets/SuperMultiplayerShooter/Scripts/PubNubUtilities.cs [hunk 3]
```diff
@@ -306,18 +317,18 @@ public void UpdatePlayerPosition(PubNub pubnub, int playerID, Vector2 position,
             movementData[5] = velocity.y;
 
             string channelName = playerPositionChannelPrefix + playerID;
-            pubnub.Signal().Message(movementData).Channel(channelName).Async((result, status) =>
+            pubnub.Signal().Message(movementData).Channel(channelName).Execute((result, status) =>
```
Comment (darryncampbell-pubnub) [inline Assets/SuperMultiplayerShooter/Scripts/PubNubUtilities.cs:320]: This makes sense, I was going to say we don't need to `await` for any of the signals but looks like you didn't do that anyway :)

File: Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs [hunk 4]
```diff
@@ -163,123 +302,198 @@ void Update()
         }
 
         /// <summary>
-        /// Called whenever the scene or game ends. Unsubscribe from all channels and channel groups to trigger events immediately.
+        /// Called whenever the scene or game ends. Unsubscribe event listeners.
         /// </summary>
         void OnDestroy()
         {
-            _pubnub.UnsubscribeAll()
```
Comment (darryncampbell-pubnub) [inline Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs:170]: Is it possible to unsubscribe from all channels as part of the shutdown process?  I tried this with some of my channels in OnDestroy() with the old SDK (as well as in the Unity ApplicationWantsToQuit() callback) but I couldn't do it as the old SDK didn't support await.

File: Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs [hunk 5]
```diff
@@ -163,123 +302,198 @@ void Update()
...
+            //Clear out the cached players when changing scenes. The list needs to be updated when returning to the scene in case
+            //there are new players.
+            PNManager.pubnubInstance.CachedPlayers.Clear();
+        }
+
+        /// <summary>
+        /// Get the User Metadata given the UserId.
+        /// </summary>
+        /// <param name="Uuid">UserId of the Player</param>
+        private async void GetUserMetadata(string Uuid)
```
Comment (darryncampbell-pubnub) [inline Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs:322]: Do you have a method I can call that essentially does:
1. If the username is cached, return the cached value
2. If the username is NOT cached, obtain the actual value and return that?

File: Assets/SuperMultiplayerShooter/Scripts/PNManager.cs (hunk 1 above)
Comment (olivercarson-pubnub) [reply Assets/SuperMultiplayerShooter/Scripts/PNManager.cs:15]: Sounds great! I've been using `pubnub.GetCurrentUserId()` and using that to compare to the cachedPlayers uuid to check if it's the player themselves.

File: Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs (hunk 5 above)
Comment (olivercarson-pubnub) [reply Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs:322]: There is no method for it, but the closest thing is via the OnPnPresence:
```
 //When user joins, check their UUID in cached players to determine if they are a new player.                 
  if (!PNManager.pubnubInstance.CachedPlayers.ContainsKey(result.Uuid))
  {
      GetUserMetadata(result.Uuid);
  }
```

Would you like me to create a function that does that?

Comment (olivercarson-pubnub) [reply Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs:345]: Added helper function in [commit 657d070](https://github.com/PubNubDevelopers/unity-multiplayer-shooter/pull/3/commits/657d070c41735530e196769610186fb7f0bff9ff).

File: Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs (hunk 4 above)
Comment (olivercarson-pubnub) [reply Assets/SuperMultiplayerShooter/Scripts/SampleMainMenu.cs:170]: Implemented null reference check in this [commit](https://github.com/PubNubDevelopers/unity-multiplayer-shooter/pull/3/commits/657d070c41735530e196769610186fb7f0bff9ff) to ensure no null reference errors occur when trying to unsubscribe from all channels.

**Review by darryncampbell-pubnub — APPROVED** [review]
Approving based on our offline conversations

**Pull request merged**